# NOAA Analysis

In [130]:
import pandas as pd
pd.set_option('display.float_format', lambda x: '%.3f' % x)
%matplotlib inline

In [47]:
event_details = pd.read_csv("results/combined_0.csv", low_memory=False);
event_details.shape

(1457423, 51)

In [7]:
event_details.head(3)

BEGIN_YEARMONTH  BEGIN_DAY  BEGIN_TIME  END_YEARMONTH  END_DAY  END_TIME  \
0           195004         28        1445         195004       28      1445   
1           195004         29        1530         195004       29      1530   
2           195007          5        1800         195007        5      1800   

   EPISODE_ID  EVENT_ID         STATE  STATE_FIPS     ...      END_RANGE  \
0         NaN  10096222      OKLAHOMA        40.0     ...            0.0   
1         NaN  10120412         TEXAS        48.0     ...            0.0   
2         NaN  10104927  PENNSYLVANIA        42.0     ...            0.0   

  END_AZIMUTH END_LOCATION BEGIN_LAT  BEGIN_LON END_LAT END_LON  \
0         NaN          NaN     35.12      -99.2   35.17  -99.20   
1         NaN          NaN     31.90      -98.6   31.73  -98.60   
2         NaN          NaN     40.58      -75.7   40.65  -75.47   

  EPISODE_NARRATIVE EVENT_NARRATIVE DATA_SOURCE  
0               NaN             NaN         PUB  
1               NaN             NaN         PUB  
2               NaN             NaN         PUB  

[3 rows x 51 columns]

### Problem 1 - summarize the property damage, by state, by year

In [57]:
event_details_state['DAMAGE_PROPERTY'].head(3)

0    250K
1     25K
2     25K
Name: DAMAGE_PROPERTY, dtype: object

The property damage column is not in number form. A quick glance shows that this field is suffixed with K representing thousand and M representing million. Next, we will convert this into number form.

In [64]:
# stolen from https://stackoverflow.com/questions/39684548/convert-the-string-2-90k-to-2900-or-5-2m-to-5200000-in-pandas-dataframe
def value_to_float(x):
    x = str(x)
    if 'K' in x:
        if len(x) > 1:
            return float(x.replace('K', '')) * 1000
        return 1000.0
    if 'M' in x:
        if len(x) > 1:
            return float(x.replace('M', '')) * 1000000
        return 1000000.0
    if 'B' in x:
        return float(x.replace('B', '')) * 1000000000
    return 0.0

250000.0

In [173]:
event_details['DAMAGE_PROPERTY_DOLLARS'] = event_details['DAMAGE_PROPERTY'].apply(value_to_float)

In [72]:
event_details['STATE'].unique()

array(['OKLAHOMA', 'TEXAS', 'PENNSYLVANIA', 'NEBRASKA', 'MISSISSIPPI',
       'NEW MEXICO', 'ARKANSAS', 'MISSOURI', 'CONNECTICUT', 'FLORIDA',
       'NORTH CAROLINA', 'ALABAMA', 'KENTUCKY', 'MARYLAND', 'MINNESOTA',
       'SOUTH DAKOTA', 'IOWA', 'LOUISIANA', 'OHIO', 'KANSAS',
       'NORTH DAKOTA', 'INDIANA', 'COLORADO', 'SOUTH CAROLINA',
       'WEST VIRGINIA', 'WYOMING', 'GEORGIA', 'WISCONSIN', 'ILLINOIS',
       'TENNESSEE', 'NEW JERSEY', 'MICHIGAN', 'CALIFORNIA',
       'MASSACHUSETTS', 'NEW HAMPSHIRE', 'OREGON', 'VIRGINIA', 'ARIZONA',
       'NEW YORK', 'MONTANA', 'MAINE', 'VERMONT', 'UTAH', 'DELAWARE',
       'IDAHO', 'WASHINGTON', 'HAWAII', 'RHODE ISLAND', 'NEVADA',
       'PUERTO RICO', 'DISTRICT OF COLUMBIA', 'ALASKA', 'VIRGIN ISLANDS',
       'GUAM', 'AMERICAN SAMOA', 'ATLANTIC SOUTH', 'LAKE ERIE',
       'LAKE HURON', 'GULF OF MEXICO', 'LAKE ST CLAIR', 'E PACIFIC',
       'HAWAII WATERS', 'LAKE MICHIGAN', 'ATLANTIC NORTH', 'LAKE ONTARIO',
       'ST LAWRENCE R', 'LAKE SUPERI

We see that there are non-states in our dataset, we need to be sure to filter these out.

In [279]:
states = pd.read_csv("states.csv")
#upper all states so they match file
states['State'] = states['State'].apply(lambda x: x.upper())
event_details_state = event_details[event_details['STATE'].isin(states['State'])].copy()
#event_details_state['DAMAGE_PROPERTY_DOLLARS'] = event_details_state['DAMAGE_PROPERTY_DOLLARS'].astype('float')
event_details_state.shape

BEGIN_YEARMONTH  BEGIN_DAY  BEGIN_TIME  END_YEARMONTH  END_DAY  END_TIME  \
0           195004         28        1445         195004       28      1445   
1           195004         29        1530         195004       29      1530   
2           195007          5        1800         195007        5      1800   
3           195007          5        1830         195007        5      1830   
4           195007         24        1440         195007       24      1440   

   EPISODE_ID  EVENT_ID         STATE  STATE_FIPS           ...            \
0         nan  10096222      OKLAHOMA      40.000           ...             
1         nan  10120412         TEXAS      48.000           ...             
2         nan  10104927  PENNSYLVANIA      42.000           ...             
3         nan  10104928  PENNSYLVANIA      42.000           ...             
4         nan  10104929  PENNSYLVANIA      42.000           ...             

   END_AZIMUTH END_LOCATION BEGIN_LAT BEGIN_LON  END_LAT END_LON  \
0          NaN          NaN    35.120   -99.200   35.170 -99.200   
1          NaN          NaN    31.900   -98.600   31.730 -98.600   
2          NaN          NaN    40.580   -75.700   40.650 -75.470   
3          NaN          NaN    40.600   -76.750      nan     nan   
4          NaN          NaN    41.630   -79.680      nan     nan   

  EPISODE_NARRATIVE EVENT_NARRATIVE DATA_SOURCE DAMAGE_PROPERTY_DOLLARS  
0               NaN             NaN         PUB              250000.000  
1               NaN             NaN         PUB               25000.000  
2               NaN             NaN         PUB               25000.000  
3               NaN             NaN         PUB                2500.000  
4               NaN             NaN         PUB                2500.000  

[5 rows x 52 columns]

We went from 1,457,423 to 1,424,943 after filtering non US states.

In [100]:
event_details_grouped = event_details_state.groupby(['STATE','YEAR'])['DAMAGE_PROPERTY_DOLLARS'].sum()

In [105]:
event_details_grouped.head()

STATE    YEAR
ALABAMA  1950     27500.000
         1951     35000.000
         1952   5452500.000
         1953   3070000.000
         1954    607530.000
Name: DAMAGE_PROPERTY_DOLLARS, dtype: float64

In [104]:
# outputs to csv file
event_details_grouped.to_csv("property_damage_summary.csv")

### Problem 2 - where to build to minimize property damage

In [280]:
event_details_grouped = event_details_state.groupby(['STATE'])['DAMAGE_PROPERTY_DOLLARS'].sum().sort_values(ascending=False)
event_details_grouped.head()

STATE
LOUISIANA     68120714540.000
TEXAS         47104718850.000
FLORIDA       41441290320.000
MISSISSIPPI   31003482190.000
NEW JERSEY    27795146590.000
Name: DAMAGE_PROPERTY_DOLLARS, dtype: float64

In terms of dollars, Louisiana is clearly in the worst shape due to disasters with over 68 billion in damages, a good portion of that due to Katrina. Unforunately, New Jersey is also in the top 5 with almost almost 28 billion in damages, which mostly can be attributed to Sandy. 



In [268]:
state_land_area = pd.read_csv('states_landarea.csv')
state_land_area.set_index('STATE', inplace=True,)
# convert to dict for constant time access
state_land_area_dict = state_land_area.to_dict()['AREA']

def calc_damage_per_sqmi(state_row):
    try:
        
        land_area = int(state_land_area_dict[state_row['STATE']].replace(',', ''))
        damage_dollars = state_row['DAMAGE_PROPERTY_DOLLARS']
        return damage_dollars/land_area
    except:
        print("Could not locate state {} in land area file".format(state))
        return None
    
    

In [269]:
event_details_state['DAMAGE_PROPERTY_DOLLARS_SQMI'] = event_details_state[['DAMAGE_PROPERTY_DOLLARS','STATE']].apply(calc_damage_per_sqmi,axis=1);
event_details_state.head()


BEGIN_YEARMONTH  BEGIN_DAY  BEGIN_TIME  END_YEARMONTH  END_DAY  END_TIME  \
0           195004         28        1445         195004       28      1445   
1           195004         29        1530         195004       29      1530   
2           195007          5        1800         195007        5      1800   
3           195007          5        1830         195007        5      1830   
4           195007         24        1440         195007       24      1440   

   EPISODE_ID  EVENT_ID         STATE  STATE_FIPS  \
0         nan  10096222      OKLAHOMA      40.000   
1         nan  10120412         TEXAS      48.000   
2         nan  10104927  PENNSYLVANIA      42.000   
3         nan  10104928  PENNSYLVANIA      42.000   
4         nan  10104929  PENNSYLVANIA      42.000   

              ...               BEGIN_LAT BEGIN_LON END_LAT END_LON  \
0             ...                  35.120   -99.200  35.170 -99.200   
1             ...                  31.900   -98.600  31.730 -98.600   
2             ...                  40.580   -75.700  40.650 -75.470   
3             ...                  40.600   -76.750     nan     nan   
4             ...                  41.630   -79.680     nan     nan   

   EPISODE_NARRATIVE EVENT_NARRATIVE DATA_SOURCE DAMAGE_PROPERTY_DOLLARS  \
0                NaN             NaN         PUB              250000.000   
1                NaN             NaN         PUB               25000.000   
2                NaN             NaN         PUB               25000.000   
3                NaN             NaN         PUB                2500.000   
4                NaN             NaN         PUB                2500.000   

  PROPERTY_DAMAGE_PER_SQMI DAMAGE_PROPERTY_DOLLARS_SQMI  
0                      NaN                        3.577  
1                      NaN                        0.093  
2                      NaN                        0.543  
3                      NaN                        0.054  
4                      NaN                        0.054  

[5 rows x 54 columns]

In [277]:
event_details_state_grouped = event_details_state.groupby(['STATE'])[['DAMAGE_PROPERTY_DOLLARS_SQMI','DAMAGE_PROPERTY_DOLLARS']].sum()

In [ ]:
# grouped by state and year
#event_details_state_grouped = event_details_state.groupby(['STATE','YEAR'])[['DAMAGE_PROPERTY_DOLLARS_SQMI','DAMAGE_PROPERTY_DOLLARS']].sum()

#####

In [278]:
event_details_state_grouped.sort_values(by = 'DAMAGE_PROPERTY_DOLLARS_SQMI',ascending=False, inplace=True)
event_details_state_grouped.head(55)

DAMAGE_PROPERTY_DOLLARS_SQMI  DAMAGE_PROPERTY_DOLLARS
STATE                                                                      
NEW JERSEY                             3186420.565          27795146590.000
DISTRICT OF COLUMBIA                   2319089.706            157698100.000
LOUISIANA                              1300559.673          68120714540.000
MISSISSIPPI                             640144.578          31003482190.000
FLORIDA                                 630209.105          41441290320.000
ALABAMA                                 229473.829          12029018120.000
NORTH CAROLINA                          176403.580           9493864260.000
TEXAS                                   175373.866          47104718850.000
OHIO                                    167318.760           7500230740.000
VERMONT                                 159491.088           1533666300.000
TENNESSEE                               158326.686           6672519870.000
CONNECTICUT                             139019.659            770585970.000
OKLAHOMA                                124856.040           8727312320.000
MISSOURI                                123881.754           8635425430.000
INDIANA                                 116346.570           4237342090.000
MASSACHUSETTS                           107663.929           1136285110.000
ARKANSAS                                105827.775           5627815270.000
IOWA                                    102367.966           5760552530.000
PENNSYLVANIA                            101408.190           4670252760.000
MARYLAND                                100401.475           1245580700.000
ILLINOIS                                 93991.402           5443418030.000
GEORGIA                                  90012.826           5349012190.000
NEW YORK                                 89589.144           4887535750.000
RHODE ISLAND                             84030.227            129826700.000
KENTUCKY                                 82836.889           3347273030.000
NORTH DAKOTA                             75326.389           5325425070.000
MINNESOTA                                64516.938           5608844560.000
DELAWARE                                 63559.293            158199080.000
COLORADO                                 62391.380           6494568290.000
KANSAS                                   61730.580           5079068660.000
MICHIGAN                                 55190.914           5337734050.000
CALIFORNIA                               53438.382           8747595900.000
WISCONSIN                                52338.958           3427992380.000
VIRGINIA                                 51383.126           2197913210.000
WEST VIRGINIA                            51124.933           1238757120.000
NEBRASKA                                 50549.714           3909919280.000
SOUTH CAROLINA                           43518.882           1393474590.000
ARIZONA                                  34565.670           3940140680.000
NEW HAMPSHIRE                            28356.063            265100830.000
NEW MEXICO                               19289.256           2345380590.000
WASHINGTON                               18342.370           1307774270.000
HAWAII                                   17723.862            193757260.000
MAINE                                    15115.223            534776600.000
OREGON                                   11583.044           1139528330.000
SOUTH DAKOTA                             10532.866            812252500.000
UTAH                                     10448.608            887055470.000
IDAHO                                    10079.622            842343910.000
NEVADA                                    9493.518           1049717260.000
WYOMING                                   2561.852            250582410.000
MONTANA                                   2160.409            317666500.000
ALASKA                                     544.972            36261550